src/lib/tuning/tuner.hpp

```diff
@@ -0,0 +1,155 @@
+#pragma once
+
+#include <chrono>
+#include <limits>
+#include <memory>
+#include <string>
+
+#include "scheduler/abstract_task.hpp"
+#include "tuning/abstract_evaluator.hpp"
+#include "tuning/abstract_selector.hpp"
+
+namespace opossum {
+
+/**
+ * A Tuner encapsulates the process of analyzing the current system state and
+ * performing specific modifications to optimize the systems performance.
+ *
+ * It uses AbstractEvaluators to generate TuningChoices which represent possible
+ * modifications together with their expected performance impact and costs.
+ * TuningChoices are transformed into a concrete operation sequence
+ * by an AbstractSelector, that also considers a cost budget that may not be
+ * exceeded at any point in the sequence.
+ *
+ * While executing the operation sequence, the Tuner considers a time budget
+ * and stops the execution once it exceeds that budget.
+ *
+ * Both the time and cost budget (the latter being enforced by the AbstractSelector)
+ * are initially infinite (i.e. disabled).
+ */
+class Tuner {
```

Comment:
Can we make any member variables const?

Reply:
Maintaining the re-configurability of the Tuner as discussed above, none of the member variables can be made const. However if const-ness would provide performance benefits, this would be a strong argument against re-configurability.

Reply:
Const-ness should not be a performance issue here